module/dkg/broker.go

```diff
@@ -166,27 +192,78 @@ func (b *Broker) Broadcast(data []byte) {
 		if err != nil {
-			b.log.Error().Err(err).Msg("failed to broadcast message")
+			log.Error().Err(err).Msgf("failed to broadcast message after %d attempts", attempts)
+			return
+		}
+		log.Info().Msgf("dkg broadcast successfully on attempt %d", attempts)
+	})
+}
+
+// SubmitResult publishes the result of the DKG protocol to the smart contract.
+func (b *Broker) SubmitResult(groupKey crypto.PublicKey, pubKeys []crypto.PublicKey) error {
+
+	// If the DKG failed locally, we will get a nil key vector here. We need to convert
+	// the nil slice to a slice of nil keys before submission.
+	//
+	// In general, if pubKeys does not have one key per participant, we cannot submit
+	// a valid result - therefore we submit a nil vector (indicating that we have
+	// completed the process, but we know that we don't have a valid result).
+	if len(pubKeys) != len(b.committee) {
+		b.log.Warn().Msgf("submitting dkg result with incomplete key vector (len=%d, expected=%d)", len(pubKeys), len(b.committee))
+		// create a key vector with one nil entry for each committee member
+		pubKeys = make([]crypto.PublicKey, len(b.committee))
+	}
+
+	backoff, err := retry.NewExponential(b.config.RetryInitialWait)
+	if err != nil {
+		b.log.Fatal().Err(err).Msg("failed to create retry mechanism")
+	}
+	backoff = retry.WithMaxRetries(b.config.PublishMaxRetries, backoff)
+	backoff = retry.WithJitterPercent(b.config.RetryJitterPct, backoff)
+
+	clientIndex, dkgContractClient := b.getInitialContractClient()
+	onMaxConsecutiveRetries := func(totalAttempts int) {
+		clientIndex, dkgContractClient = b.updateContractClient(clientIndex)
+		b.log.Warn().Msgf("submit result: retrying on attempt (%d) with fallback access node at index (%d)", totalAttempts, clientIndex)
+	}
+	backoff = retrymiddleware.AfterConsecutiveFailures(b.config.RetryMaxConsecutiveFailures, backoff, onMaxConsecutiveRetries)
+
+	attempts := 1
+	err = retry.Do(b.unit.Ctx(), backoff, func(ctx context.Context) error {
+		err := dkgContractClient.SubmitResult(groupKey, pubKeys)
+		if err != nil {
+			b.log.Error().Err(err).Msgf("error submitting DKG result, retrying (attempt %d)", attempts)
+			attempts++
+			return retry.RetryableError(err)
 		}
+
+		// update our last successful client index for future calls
+		b.updateLastSuccessfulClient(clientIndex)
+		return nil
 	})
+	if err != nil {
+		return fmt.Errorf("failed to submit dkg result after %d attempts: %w", attempts, err)
+	}
+
+	b.log.Info().Msgf("dkg result submitted successfully on attempt %d", attempts)
+	return nil
 }
 
 // Disqualify flags that a node is misbehaving and got disqualified
 func (b *Broker) Disqualify(node int, log string) {
-	// The Warn log is used by the integration tests to check if this method is
-	// called.
+	// The Warn log is used by the integration tests to check if this method is called.
 	b.log.Warn().Msgf("participant %d is disqualifying participant %d because: %s", b.myIndex, node, log)
```

Comment:
how to know the node id by the index? is it worth to include the node ID in the log?

Reply:
Changed to log both index and node ID: https://github.com/onflow/flow-go/pull/1950/commits/9c084fec9bfedfd7cc05df8ae06f7dca86bb8b5a